src/core/ddsi/src/q_entity.c

```diff
@@ -3894,6 +4209,132 @@ static void add_proxy_builtin_endpoints(
+  if (proxypp) {
+    pp = ephash_lookup_participant_guid (gv->guid_hash, lpguid);
+    if (pp) {
+      switch(result)
+      {
+      case STATE_HANDSHAKE_PROCESSED:
+        shared_secret = ddsi_handshake_get_shared_secret(handshake);
+        DDS_CLOG (DDS_LC_DISCOVERY, &gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") processed\n", PGUID (*lpguid), PGUID (*ppguid));
+        permissions_hdl = q_omg_security_check_remote_participant_permissions(gv->config.domainId, pp, proxypp);
+        if (permissions_hdl != 0) {
+          q_omg_security_register_remote_participant(pp, proxypp, shared_secret, permissions_hdl);
+          match_volatile_secure_endpoints(pp, proxypp);
+        }
+        break;
+
+      case STATE_HANDSHAKE_SEND_TOKENS:
+        DDS_CLOG (DDS_LC_DISCOVERY, &gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") send tokens\n", PGUID (*lpguid), PGUID (*ppguid));
+        q_omg_security_participant_send_tokens(pp, proxypp);
+        break;
+
+      case STATE_HANDSHAKE_OK:
+        DDS_CLOG (DDS_LC_DISCOVERY, &gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") succeeded\n", PGUID (*lpguid), PGUID (*ppguid));
+        update_proxy_participant_endpoint_matching(proxypp, pp);
+        ddsi_hsadmin_remove_from_fsm(proxypp->handshake_admin, handshake);
+        break;
+
+      case STATE_HANDSHAKE_TIMED_OUT:
+        DDS_CERROR (&gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") failed: (%d) Timed out\n", PGUID (*lpguid), PGUID (*ppguid), (int)result);
+        if (q_omg_participant_allow_unauthenticated(pp)) {
+          downgrade_to_nonsecure(proxypp);
+          update_proxy_participant_endpoint_matching(proxypp, pp);
+        }
+        ddsi_hsadmin_remove_from_fsm(proxypp->handshake_admin, handshake);
+        break;
+      case STATE_HANDSHAKE_FAILED:
+        DDS_CERROR (&gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") failed: (%d) Failed\n", PGUID (*lpguid), PGUID (*ppguid), (int)result);
+        if (q_omg_participant_allow_unauthenticated(pp)) {
+          downgrade_to_nonsecure(proxypp);
+          update_proxy_participant_endpoint_matching(proxypp, pp);
+        }
+        ddsi_hsadmin_remove_from_fsm(proxypp->handshake_admin, handshake);
+        break;
+      default:
+        DDS_CERROR (&gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") failed: (%d) Unknown failure\n", PGUID (*lpguid), PGUID (*ppguid), (int)result);
+        ddsi_hsadmin_remove_from_fsm(proxypp->handshake_admin, handshake);
+        break;
+      }
+    }
+  }
+}
+
+static int proxy_participant_check_security_info(struct q_globals *gv, struct proxy_participant *proxypp)
+{
+  int r = 1;
+  struct participant *pp;
+  struct ephash_enum_participant est;
+
+  ephash_enum_participant_init (&est, gv->guid_hash);
+  while (((pp = ephash_enum_participant_next (&est)) != NULL) && (r == 1)) {
+    if (!q_omg_is_similar_participant_security_info(pp, proxypp)) {
```

Comment:
I have to think about this ... does it make sense to reject a proxy participant if there is a mismatch with a single participant? Should it not be rejected only when there is a mismatch with all participants?

The way the stack handles multiple participants, I can see that this may be necessary. That would be an argument to allow multiple instantiations of a single domain in a single process (which is something outside the scope of this PR, and of other security-related PRs).

Reply:
Yes I also think that it is better to reject the proxy participant only when there is a mismatch with all the participants. In the current implementation there could be only one security configuration thus all participants would have the same security settings.